### DISCUSSION TITLE
Behavior of the default rules for object with index

### DISCUSSION THREAD

**Discussion by mingweishih** [open-policy-agent/community#339] [OPA and Rego]
When trying the default rules, I notice that `default var[index] = value` has the following behaviors
- If the index is a scalar value, OPA will initialize the var as an object.
  For example, 
  Input
  ```
  default z[1] = 1
  ```
  Output
  ```
  "z": { "1": 1 }
  ```
- If the index is an undefined reference, the OPA will raise error
  Input
  ```
  default m[n] = 1
  ```
  Output
  ```
  rego_unsafe_var_error: var n is unsafe
  ```
- If the index is defined reference or a comprehension, OPA will ignore the index
  Case 1:
  Input
  ```
  n = 5
  default m[n] = 1
  ```
  Output
  ```
  "m": 1,
  "n": 5
  ```
  Case 2:
  Input
  ```
  default y[[x | x = 5]] = 1
  ```
  Output
  ```
  "y": 1
  ```

Are these behaviors by-design?

**Comment by srenatus** [comment]
Case 1 and 2 look like bugs to me. They should both be invalid and rejected by the compiler.

**Reply by anderseknert** [reply]
Wow! Multiple issues reported with default assignment on one day. And as far as I know the _only_ reported issues on default assignment 😅 

https://github.com/orgs/open-policy-agent/discussions/338
https://github.com/open-policy-agent/opa/issues/5645

**Reply by mingweishih** [reply]
> Case 1 and 2 look like bugs to me. They should both be invalid and rejected by the compiler.

So, is the default rule supposed to only allow variable assignment?
